### ISSUE TITLE
New room list: large account performance

### ISSUE THREAD

**Issue by turt2live** [vector-im/riot-web#14035]
Follow on from https://github.com/vector-im/riot-web/issues/14034 - make it not noticeable for large accounts.

**Comment by turt2live** [comment]
With the above PRs merged, performance is looking a lot better:
![image](https://user-images.githubusercontent.com/1190097/87103808-4110ab00-c213-11ea-92f9-abbb871d0104.png)

At 37.77ms per operation, most of which is 1-2ms function calls, we're doing pretty good. Incremental updates are well down from 100ms to just 3.15ms on average (previous metrics: https://github.com/matrix-org/matrix-react-sdk/pull/4253), however that's still acceptable. The remaining ~35ms are spent preparing for a render of the update which just took place.

We do still fire a boatload of updates though (as indicated by the flamegraph, which is only a tens of seconds of time).

When we disable all the logging, incremental generation falls by nearly a full millisecond.

Initial generation takes just 3.5ms on my account, without logging. With logging it's a bit longer but not by much.

Most of the rendering time appears to be in the depths of React - we have some control over this but overall the benefit of going into it much further isn't totally worth it at this point. 

Although the numbers tell an amazing story, we're leaving this issue open until others with large accounts (Matthew) are happy with the overall performance. It's not expected to be perfect, but should be much closer to perfect than unusable.

**Comment by turt2live** [comment]
for comparison: the old room list achieved 394ms initial generation and 1.5ms incremental, with the list before that getting 0.13ms initial and 0.26ms incremental.

The old old room list (~2018) got such great speeds because it wasn't doing anything terribly complicated, but with the "breadcrumb" (now called importance) algorithm there were a number of complex features introduced like sticky rooms and advanced sorting behaviours. When comparing the old room list versus this new one, the initial generation time is much better because we're tracking rooms differently - this is also what consequently increases the incremental time. We spend slightly more time making sure we're going to put a room in the right spot to make subsequent updates `O(n)` instead of exponential or logarithmic. 

If we remove some safety code, we could probably shave a millisecond or two off the incremental time, however at a cost of a couple milliseconds I think it's worth keeping it in.

**Comment by turt2live** [comment]
This got the seal of approval I was looking for: "it feels better".

**Comment by Half-Shot** [comment]
It's really not feeling better from my side. Admittedly, I am not running the latest and greatest hardware but I don't believe we are targeting those users anyway. I'll gather logs

**Comment by jcgruenhage** [comment]
I've switched to the new room list, my account is okay on my workstation (I didn't notice issues there), but it turned element on my Pinebook Pro from "okay" to "infuriatingly slow".

In the configuration I'd want to use ideally (all rooms, top 10 of other categories), it takes 8 seconds to switch a room, 2 seconds to even highlight it when hovering over it with the mouse. Composing messages is a pain to, since the composer lags between a few hundred ms to a few s behind what I'm actually typing. It's slowing down my laptop enough, that I actually closed element to submit this comment...
